### ISSUE TITLE
Rejigger Generic1

### ISSUE THREAD

**Issue by treeowl** [phadej/staged-gg#6]
It's [definitely possible](https://gist.github.com/treeowl/7cddf019b7154b08987e1024959103ec) to calculate the `Rep1` of a type from its `GHC.Generics.Rep` instead of its `GHC.Generics.Rep1`. This seems better, since `GHC.Generics.Generic1` is kind of broken. That would, of course, be a breaking change, but I doubt this package has many users.

The next interesting question is whether it's possible to implement `Generic1` *methods* using generics. I bet it is, and I bet it can even be done without hideous amounts of code duplication.

**Comment by phadej** [comment]
My offer for you to take over this package is still valid.

**Comment by phadej** [comment]
TBH, i think this package has zero users, even I don't use it myself. So I wouldn't be worried about breaking stuff.

I don't understand what you mean by GHC.Generics.Generic1 being broken, why it is not fixed?

**Comment by treeowl** [comment]
The right-associative composition was a mistake, not fixed for backwards compatibility. Tiresome.

**Comment by phadej** [comment]
But `staged-gg` "fixes" that, doesn't it?

**Comment by treeowl** [comment]
Yes, but the default definition of Rep1 is built off `GHCG.Rep1`, which only exists if there are enough `Functor` instances for GHC to derive `Generic1`.

**Comment by phadej** [comment]
Well... I'm mixed.

There are plenty of evidence that GHC.Generics should be fixed: no instances, doesn't generalize to staged or linear.. but somewhat courner-casy.

But these aren't fixed because of some backward compat, which practioners (you & me) don't really care that much.

Can one derive "right" `GHC.Generic1` from current `GHC.Generic`? I'll accept that roundabout (as then there is *some* hope it would be adopted to GHC proper) but not the edirect short-cut.

**Comment by treeowl** [comment]
I linked above to my implementation of left-handed `Generic1` from `Generic`. Basically,

```
Generic1 f ~= (forall a. Generic (f a)) /\ the last parameter is only ever a final argument
```
